### DISCUSSION TITLE
Context-aware tokens / utility classes

### DISCUSSION THREAD

**Discussion by janhassel** [carbon-design-system/carbon#7701] [Ideas]
Related: https://github.com/carbon-design-system/carbon/issues/4358#issuecomment-554778016, https://github.com/carbon-design-system/carbon/issues/2912, https://github.com/carbon-design-system/carbon/issues/4826, https://github.com/carbon-design-system/carbon/pull/5598, https://github.com/carbon-design-system/carbon/pull/5914

When using tiles / cards / containers to layout an interface, it quickly becomes challenging to apply the correct theming to all elements due to the overlaps in some color tokens. Having a container that uses `ui-01` as its background requires all inputs inside to use `props.light` in order to switch their background color from `field-01` to `field-02`. Support for `props.light` is not given for all components though. The DataTable for example does not offer this prop and can therefore not be correctly displayed within a `ui-01` container.

One solution to this is to apply different themes to different zones. One could create a `g10/ui-01` container and then emit the `white` theme to its content, effectively treating the inside of the container as it would be a page using `white`. This would correctly render all components without using `props.light` and therefore supports the DataTable. However, this solution can also become tricky when nesting more objects and adds a lot of duplicates to the code.

From a developer experience standpoint it would be great if the components just "knew" how to render properly and in which context they are. One approach could be to introduce utility classes (@joshblack I believe you were exploring this direction before) that would emit only the necessary updated color tokens for their content.

---

**Example**

An input component would no longer use `field-01` as its background and have a light variant to switch to `field-02` but it would rather rely on a context-aware token `field`. By default, `field` could be set to `field-01`.
To make a container, developers would use a utility class, let's call it `cds-bg-ui`. This class would set the container's background color to `ui` (also context-aware) and override relevant color tokens for its content.
In this example, we'd want to use `field-02` for all inputs. Any nested container should then use `ui-02` as their background and revert to `field-01` for its inputs.

```css
:root {
  --cds-ui-01: #ffffff;
  --cds-ui-02: #f4f4f4;
  --cds-field-01: #ffffff;
  --cds-field-02: #f4f4f4;

  --cds-ui: var(--cds-ui-01);
  --cds-field: var(--cds-field-01);
}

.cds-bg-ui {
  background-color: var(--cds-ui);
  
  --cds-field: var(--cds-field-02);
}

.cds-bg-ui .cds-bg-ui {
  --cds-ui: var(--cds-ui-02);
  --cds-field: var(--cds-field-01);
}
```
_(Since there are only ui-01 and ui-02 that describe a container background in the color usage guidelines, I assume it's a design choice to not encourage more deeply nested levels)_

A developer would then not need to care about where components are rendered and if they might need any additional props.

```html
<div class="cds-bg-ui">
  <TextInput />

  <div class="cds-bg-ui">
    <TextInput />
  </div>
</div>

```

Quick codepen example applying above described behaviour: https://codepen.io/janhassel/pen/7f593afc5cac9c7e9f94f506850e54b1?editors=1100

--- 

I think an approach like this could vastly improve the developer experience and make sure colors are used as intended (avoid having field colors and background colors the same or wrong usage of zoned themes). I have to say that I didn't work a lot with css frameworks like tailwind and this idea of context-awareness might not fit into the "only-one-rule-and-nothing-more" mindset. It would definitely make the styling more opinionated, although that might be a good thing for a design system (to ensure correct usage).

Would love y'all's thoughts on this.

**Comment by joshblack** [comment]
@janhassel love this exploration!

Zones/inline themes are definitely an artifact of the Duo exploration days that we never were really able to tackle in v10. I think that, unfortunately, some designs may still use this idea of "themes in themes" and the resulting compromise was the `light` prop, as you noted, with the limitations described above.

Ultimately you're totally right, CSS Custom Properties in this space would be ideal. It would allow us to solve this cleanly and efficiently and would require no re-write of the components themselves. It would just work. The mechanism for this may end up just being theme classes, and for the developer that may end up looking like:

```jsx
<Theme value="white">
  <PageSection />
  <Theme value="g10">
    <PageSection />
  </Theme>
  <PageSection />
</Theme>
```

Or more simply the following theme classes: `bx--white`, `bx--g10`, `bx--g90`, `bx--g100` that would correctly define CSS Custom Properties for children components to inherit from.

Curious what you think of this @janhassel and if anything is missing based on the use-cases you described above.

**Reply by janhassel** [reply]
@joshblack The theme-class approach you mention is actually how I've structured my prototypes for a while now with the only difference being using a semantic naming (something like `bx--outer` and `bx--inner`) instead of the absolute theme names to account for light and dark mode:

|  | `bx--outer` | `bx--inner` |
| :- | :- | :- |
| Light |  `g10` | `white` |
| Dark |  `g100` | `g90` |

While this seemed to be an easy way to achieve the correct styling, there are two major issues I faced:

--- 

1. The mental model of how backgrounds "stack" differs between the light themes and dark themes. In white and g10, if a background is displayed on a background, they always alternate between white and gray. The dark themes however follow the model of always lightening the shade.

| Level | white | g10 | g90 | g100 |
| :- | :- | :- | :- | :- |
| 1 / `ui-background` | white | gray10 | gray90 | gray100 | 
| 2 / `ui-01` and `field-01` | gray10 | white | gray80 | gray90 |
| 3 / `ui-02` and `field-02` | white | gray10 | gray70 | gray80 |

Now this issue comes apparent when needing to place an input on a `ui-02` background. In the light themes it doesn't matter as it always just alternates, but in the dark themes it makes a difference whether you're using an inline theme or not.

![image](https://user-images.githubusercontent.com/28265588/106571973-1fbd3680-6538-11eb-9bad-fdc3d876650f.png)

In this mockup, the left examples use one theme and the `light` prop (or the described, context-aware token approach for that matter 😄 ) whereas the right examples use the inline themes. By emitting the g90 theme on the right bottom sub-section, the input field must use `field-02` as `field-01` and `ui-01` are the same color and the input would therefore not have any visual distinction to the background other than the bottom border. When using one theme (on the left), the input would need to be `field-01` for the same reason (`field-02` being the same `ui-02`).

In the end, this brings up the question of which approach is the correct one. Should backgrounds always stack to a ligher color in dark themes as shown on the right or should one stick to what one theme provides?

On a side note though: Using DataTables within containers with the proposed context-aware approach does not work out-of-the-box since `ui-02` and `ui-03` are the same color in the dark themes, meaning that all dividing lines would disappear from a DataTable.

---

2. The second point is not really a technical issue or difficulty but rather an aspect of developer experience. Where using theme classes still requires the developer to define which theme should be emitted on the children, a context-aware approach would remove this layer and adjust it automagically. This *can* ensure compliance to the guidelines of the design system.

Though this is definitely a point for discussion, if this "opaqueness" wouldn't cause too much confusion or introduce new limitations I'm missing right now.

**Reply by joshblack** [reply]
@janhassel totally get what you mean from the ergonomics perspective, I'm not sure what the component could be called but if we called it a generic `Section` component for now we could totally pull something off like:

```jsx
<Section>
  <p>This renders in the white theme</p>
  <Section>
    <p>This renders in the g10 theme</p>
  </Section>
</Section>
```

With an optional prop to change the behavior from white to another theme, nested sections would know what theme to swap to based on the context that it is placed in.

---

As per the correct behavior, it'd be great if we could use inline themes but I just posted this message to design to see what they say real quick.

**Reply by joshblack** [reply]
@janhassel just wanted to circle back, seems like the right-hand side is the preferred solution here btw (theme in theme)

**Reply by aagonzales** [reply]
>In the end, this brings up the question of which approach is the correct one. Should backgrounds always stack to a lighter color in dark themes as shown on the right or should one stick to what one theme provides?

The right side is the correct design intent. The color layering logic does work different between the light and dark themes.

**Reply by janhassel** [reply]
@joshblack Looking at the new token names @aagonzales shared, maybe `<Layer>` would be appropriate?
Though I'm wondering: would this be an intermediate solution until full support for CSS custom properties (aka v11) is released or would you prefer context-awareness with the help of explicit components over the use of utility classes?

@aagonzales Thanks for the clarification!

**Reply by joshblack** [reply]
@janhassel `Layer` is a tricky one since we also use it for shadows 😞  (or at least we used to). I kind of like `Zone` as a concept but not sure how inferrable it is 🤷‍♂️ 

For the release strategy, I definitely think we want this stuff enabled by default for Custom Properties but it would make sense to provide an interim step as you mentioned.

**Comment by aagonzales** [comment]
> On a side note though: Using DataTables within containers with the proposed context-aware approach does not work out-of-the-box since ui-02 and ui-03 are the same color in the dark themes, meaning that all dividing lines would disappear from a DataTable.

We have actually been working on updating our [design tokens](https://www.notion.so/carbondesignsystem/List-of-to-be-tokens-031c30ccd0e142858c79b33ad2fdc92e) for "v11." We initially were intending to build out a structure to better support the "light" variants in each theme and address some of these layering issues you have pointed out (we were calling that solution the "alt" tokens). BUT if we can implement this "inline theming" then we can rethink the new tokens with that in mind.

**Reply by janhassel** [reply]
Thanks a lot for sharing this link! I'd be very interested in the planning for future token usage and theming approaches and can probably spare a bit of time to try a few things out / explore some options. From my experience so far, color tokens and correct theming are points where I oftentimes see designers and developer struggle and I'd love to help ideate a better structure around them.

**Reply by aagonzales** [reply]
Yeah we'd love to hear any thoughts you have on them! I'm in the process of writing up an issue to explain the new proposal token set (not final yet still open to changes) and illustrate the logic. I'll share it with you when I'm done with that. 

I've heard from several people how much designers and dev struggle with understanding the tokens which was a big push making them more user friendly the next round. They are still such a new concept and every company seems to do them slightly differently. 

I think our theming structure and varying color logic also adds a layer of complexity to our tokens that make this a very interesting problem to work on. Also the fact that design tokens are harder to understand also makes it harder to get feedback on in general so again I'd love to hear anything you have thoughts on!

**Comment by jnm2377** [comment]
@janhassel not sure if you've seen this discussion #7743. It's related to the issue you bring up here, and @aagonzales proposes 2 solutions.

**Reply by janhassel** [reply]
Ah, thanks @jnm2377! Haven't been following the discussion going on over there. I'll follow up in the other thread 🙂
